src/articles/2024-01-21_DDEV-for-js.md

```diff
@@ -0,0 +1,327 @@
+  ProxyPassReverse / http://localhost:5173/
+
+  ErrorLog /dev/stdout
+  CustomLog ${APACHE_LOG_DIR}/access.log combined
+</VirtualHost>
+
+<VirtualHost *:443>
+  SSLEngine on
+  SSLCertificateFile /etc/ssl/certs/master.crt
+  SSLCertificateKeyFile /etc/ssl/certs/master.key
+
+  RewriteEngine On
+  RewriteCond %{HTTP:X-Forwarded-Proto} =https
+  RewriteCond    %{DOCUMENT_ROOT}%{REQUEST_FILENAME} -d
+  RewriteRule    ^(.+[^/])$           https://%{HTTP_HOST}$1/ [redirect,last]
+
+  SetEnvIf X-Forwarded-Proto "https" HTTPS=on
+
+  ServerName app.my-project.ddev.site
+  ProxyPreserveHost On
+  ProxyPass / http://localhost:5173/
+  ProxyPassReverse / http://localhost:5173/
+
+  ErrorLog /dev/stdout
+  CustomLog ${APACHE_LOG_DIR}/access.log combined
+</VirtualHost>
+```
+
+### Restart DDEV
+
+Once your hostnames, node version, and webserver config are all squared away, you'll need to run `ddev restart` for the changes to take effect. When command finishes, you'll see
+
+```txt
+Restarted my-project
+Your project can be reached at https://my-project.ddev.site https://api.my-project.ddev.site https://app.my-project.ddev.site
+```
+
+At this point, your new URLs will point into the webserver at the correct ports, but our apps aren't running. We could stop here and either pass commands in with `ddev exec` or tunnel in with `ddev ssh` to start the node processes inside the web container.
+
+```txt
+$ ddev ssh
+
+andy@my-project-web:/var/www/html$ cd svelte/
+andy@my-project-web:/var/www/html/svelte$ npm run dev
+```
+
+But if we run our apps this way, it's an extra step to onboard new devs with & we still have a terminal window taken over by the process. Let's boost the developer experience.
+
+## Developer Experience Improvements
+
+There are three main DX improvements we can make to make onboarding new developers as seamless as possible. First, we can utilize a process management tool like PM2 to manage our processes in the container. Second, we can make some custom commands to create easy-to-remember startup directions or to handle specific tools in the container. Third, DDEV has the ability to run commands both on the host machine and in the containers during a few key lifecycle stages.
+
+### Using PM2
+
+[PM2](https://pm2.keymetrics.io/) is an "Advanced, production process manager for Node.JS". We can use this tool to start our app without clogging up an entire terminal pane and it will also allow us to monitor the process's logs, memory usage, and some other key stats. There are a ton of options available, but we'll just use a few basics.
+
+PM2's instructions tell us to install the tool globally, so we'll handle during DDEV's build process with a custom Dockerfile in `.ddev/web-build/Dockerfile`. For more info on how DDEV handles custom Dockerfiles, see [here](https://ddev.readthedocs.io/en/stable/users/extend/customizing-images/#adding-extra-dockerfiles-for-webimage-and-dbimage)
+
+```Dockerfile
+RUN npm install -g pm2
```

Comment:
This should specify a version. Otherwise, if you need to upgrade, it won't force a rebuild on other developer machines (or even your own!).